api/prometheus/v1/api.go

```diff
@@ -169,6 +240,63 @@ type queryResult struct {
 	v model.Value
 }
 
+func (rg *RuleGroup) UnmarshalJSON(b []byte) error {
+	v := struct {
+		Name     string                   `json:"name"`
+		File     string                   `json:"file"`
+		Interval float64                  `json:"interval"`
+		Rules    []map[string]interface{} `json:"rules"`
+	}{}
+
+	if err := json.Unmarshal(b, &v); err != nil {
+		return err
+	}
+
+	rg.Name = v.Name
+	rg.File = v.File
+	rg.Interval = v.Interval
+
+	var alertingRules []AlertingRule
+	var recordingRules []RecordingRule
+
+	for _, rule := range v.Rules {
+		// Because both recording and alerting rules are stored in the same
```

Comment:
They are stored in the same JSON array as the ordering among them matters (See https://github.com/prometheus/prometheus/pull/4318#discussion_r202675025 for more details). That way users can understand which rules depend on which.

I am not sure this is a blocker. But probably worth mentioning in a comment somewhere. What do you think?

Reply:
Thanks for pointing me to that.

To keep things consistent, I've modified the client to preserve rule order as returned by the API, and added some documentation on this behavior regarding how to use a type switch to access each recording or alerting rule.